File: binary-tree-right-side-view/binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<vector<int>>ans;
        vector <int> a;
        
        if(root==NULL) return a;
        
        //queue will store node 
        queue<TreeNode*>q;
        // a.push_back(root -> val);
        q.push(root);
        while(!q.empty()){            
            int n = q.size();
            
            int v;
                
            for(int i=0;i<n;i++){
                
                TreeNode*f=q.front();
                if ( i == n-1 ) v = f -> val;
                
                q.pop();
                
                //check for left node : 
                if(f->left){
                    q.push(f->left);
                }
                
                //check for right node : 
                if(f->right){
                    q.push(f->right);
                }
            }
            a.push_back(v);
            // ans.push_back(v);
        }
        return a;
    }
};

File: binary-tree-right-side-view/binary-tree-right-side-view.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
const preOrder = function(node, arr, level) {
  if(!node) return;
  if(level >= arr.length){
    arr.push(node.val);
  }
  if(node.right) {
    preOrder(node.right, arr, level+1);
  }
  if(node.left) {
    preOrder(node.left, arr, level+1);
  }
}

const rightSideView = function (root) {
  if(!root) return [];
  let arr = [];
  preOrder(root, arr, 0);
  return arr;
};
